src/schedule/zephyr_preempt_domain.c

```diff
@@ -0,0 +1,460 @@
+	 */
+	for (i = 0; i < domain->num_dma; i++) {
+		crt_dma = domain->dma_array + i;
+
+		for (j = 0; j < crt_dma->plat_data.channels; j++) {
+			crt_chan = crt_dma->chan + j;
+			crt_data = &domain->data[i][j];
+
+			/* skip if channel is not a scheduling source */
+			if (!dma_is_scheduling_source(crt_chan))
+				continue;
+
+			/* skip if channel is not active */
+			if (crt_chan->status != COMP_STATE_ACTIVE)
+				continue;
+
+			/* skip if channel not owned by current core */
+			if (core != crt_chan->core)
+				continue;
+
+			/* skip if IRQ was already registered */
+			if (crt_data->enabled_irq)
+				continue;
+
+			/* get IRQ number for current channel */
+			irq = interrupt_get_irq(dma_chan_irq(crt_dma, j),
+						dma_chan_irq_name(crt_dma, j));
+
+			crt_data->irq = irq;
+			crt_data->channel = crt_chan;
+			crt_data->dt = dt;
+
+			if (dt->started) {
+				/* if the Zephyr thread was started, we
+				 * can safely enable the DMA IRQs
+				 */
+				ret = enable_dma_irq(crt_data);
+				if (ret < 0)
+					return ret;
+
+				crt_data->enabled_irq = true;
+			}
+
+			/* let caller know we have found a channel */
+			*data = crt_data;
+
+			return 0;
+		}
+	}
+
+	/* if this point is reached then that means we weren't able to
+	 * find a suitable channel, let caller know
+	 */
+	return -EINVAL;
+}
+
+static int zephyr_preempt_domain_register(struct ll_schedule_domain *domain,
+					  struct task *task,
+					  void (*handler)(void *arg),
+					  void *arg)
+{
+	struct zephyr_preempt_domain *zephyr_preempt_domain;
+	struct zephyr_preempt_domain_thread *dt;
+	struct pipeline_task *pipe_task;
+	struct zephyr_preempt_domain_data *data;
+	k_tid_t thread;
+	int core, ret;
+	char thread_name[] = "ll_thread0";
+
+	zephyr_preempt_domain = ll_sch_get_pdata(domain);
+	core = cpu_get_id();
+	data = NULL;
+	dt = zephyr_preempt_domain->domain_thread + core;
+	pipe_task = pipeline_task_get(task);
+
+	tr_info(&ll_tr, "zephyr_preempt_domain_register()");
+
+	/* the DMA IRQ has to be registered before the Zephyr thread is
+	 * started.
+	 *
+	 * this is done because we can have multiple DMA IRQs giving
+	 * resources to the Zephyr thread semaphore on the same core
+	 */
+	ret = register_dma_irq(zephyr_preempt_domain,
+			       &data,
+			       dt,
+			       pipe_task,
+			       core);
+
+	if (ret < 0) {
+		tr_err(&ll_tr,
+		       "failed to register DMA IRQ for pipe task %p on core %d",
+		       pipe_task, core);
+
+		return ret;
+	}
+
+	/* skip if Zephyr thread was already started on this core */
+	if (dt->handler)
+		return 0;
+
+	dt->handler = handler;
+	dt->arg = arg;
+
+	/* prepare work semaphore */
+	k_sem_init(&dt->sem, 0, SEM_LIMIT);
+
+	thread_name[sizeof(thread_name) - 2] = '0' + core;
+
+	/* create Zephyr thread */
+	thread = k_thread_create(&dt->ll_thread,
+				 zephyr_preempt_domain_stack[core],
+				 ZEPHYR_LL_STACK_SIZE,
+				 zephyr_preempt_domain_thread_fn,
+				 dt,
+				 NULL,
+				 NULL,
+				 CONFIG_NUM_PREEMPT_PRIORITIES - 1,
```

Comment:
I think we can read priority from topology. i.e. say we have 2 DMA or timer based preemptable threads we may want one to have priority over the other.

Reply:
I agree with you but for now I think it's best to leave the priority as-is since it's not really a priority to have the threads priorities based on the pipelines. Also, this would beg the following question: what priority should the thread have if we try to schedule multiple pipeline tasks having different priorities using the same thread?